Telemetry2 debugfs

jsarha opened the pull request.

Commit:
ASoC: SOF: IPC4: Add telemetry2 debug slot type

Add new debug window slot type for second telemetry data slot.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

Commit:
ASoC: SOF: IPC4: Simple debugfs file for telemetry2 debug window slot

The "telemetry2" debugfs file simply maps the telemetry2 debug slot as
a debugfs file. The decoding is left to user-space tools.

The telemetry2 debug window contents is encoded in SOF FW code found
under sof/src/debug/telemetry directory in the SOF FW sources.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

ranj063 commented on include/sound/sof/ipc4/header.h:
```diff
@@ -536,6 +536,7 @@ struct sof_ipc4_notify_resource_data {
 #define SOF_IPC4_DEBUG_SLOT_DEBUG_LOG		0x474f4c00 /* byte 0: core ID */
```
@jsarha can you please elaborate a bit on what telemetry2 is, the need for it and how it is different from the existing telemetry?

jsarha replied:
It tries to be what Liam has described here:
https://github.com/thesofproject/sof/issues/5521#issuecomment-1880945065

plbossart replied:
yeah, but no one can figure this out. You have to add a simple enough abstract so that the PR is self-contained and self-explanatory.

ranj063 commented on sound/soc/sof/ipc4-telemetry2.c:
```diff
@@ -0,0 +1,81 @@
+
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
```
@plbossart this should be GPL-2.0 only right?

plbossart replied:
Correct @ranj063. @jsarha we don't use dual-licensing when the infrastructure is specific to Linux. In this case debugfs is a Linux-specific solution which cannot be reused in other solutions, so GPL-2.0-only it is.

jsarha replied:
Ok, just copy-pasted this from somewhere. I try to find more suitable source.

plbossart commented on sound/soc/sof/ipc4-telemetry2.h:
```diff
@@ -0,0 +1,14 @@
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * Copyright(c) 2024 Intel Corporation. All rights reserved.
```
@jsarha also drop the 'All rights reserved'. We removed all this recently.

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -536,6 +536,7 @@ struct sof_ipc4_notify_resource_data {
 #define SOF_IPC4_DEBUG_SLOT_DEBUG_LOG		0x474f4c00 /* byte 0: core ID */
 #define SOF_IPC4_DEBUG_SLOT_GDB_STUB		0x42444700
 #define SOF_IPC4_DEBUG_SLOT_TELEMETRY		0x4c455400
+#define SOF_IPC4_DEBUG_SLOT_TELEMETRY2		0x4c455500
```
commit message: a 'second telemetry slot'?

so we can have two slots for telemetry used concurrently? if not, how do we know which one is valid?

That's not a good start for a review in general.

jsarha replied:
I'd assume this would require some discussion. The old telemetry slot is not very extendable so something new is needed to pass more debugging information to SOF side.

plbossart commented on sound/soc/sof/ipc4-telemetry2.c:
```diff
@@ -0,0 +1,81 @@
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2024 Intel Corporation. All rights reserved.
```
drop "All rights reserved."

plbossart commented on sound/soc/sof/ipc4-telemetry2.c:
```diff
@@ -0,0 +1,81 @@
+
+	offset = sof_ipc4_find_debug_slot_offset_by_type(sdev, type);
+	if (!offset)
+		return -EFAULT;
```
is this even remotely backwards compatible with previous generations, e.g. if you update the kernel but not the firmware?

jsarha replied:
Sure it is, as much as it can be. There of course is no previous generations of this particular feature, but its using the old debug slot reservation mechanism on the FW side and using the same function as mtrace and telemetry(1) to look for the specific slot based on the id. Naturally, there is not much we can do if the ID is not found.

plbossart replied:
the kernel should continue if it doesn't find an ID that only supported in newer versions of firmware.
I think this feature should be clearly advertised in the extended manifest, checking for the presence of an ID is a bit weird.

In this case it's not lethal because we actually ignore ALL errors from this function, so even returning -EFAULT is useless. You don't need an error code...

plbossart commented on sound/soc/sof/ipc4-telemetry2.c:
```diff
@@ -0,0 +1,81 @@
+#include "sof-priv.h"
+#include "ops.h"
+#include "ipc4-telemetry2.h"
+#include "ipc4-priv.h"
```
commit message "left to user-space tools"

What tools are you referring to? are they open-source?

jsarha replied:
The first prototype tool for extracting the thread information from the telemetry2 slot is implemented as third commit in this PR:
https://github.com/thesofproject/sof/pull/9084

But I am vague on purpose. This debugfs file just shows the debug slot contents as it is in a file, and thus the Linux driver part can be quite agnostic of the both the FW and the tools to decode the information. The tool should then check that the ABI version in the debug window is something it can understand.

plbossart replied:
At minimum you should state that the tools to extract the thread information will be open-sourced. That's a common concern for reviewers/maintainers that a new kernel interface is exposed but it can only used with private tools. In this case this is not the direction at all so an additional comment would help.

plbossart commented on sound/soc/sof/Makefile:
```diff
@@ -11,7 +11,7 @@ snd-sof-objs +=	ipc3.o ipc3-loader.o ipc3-topology.o ipc3-control.o ipc3-pcm.o\
 ifneq ($(CONFIG_SND_SOC_SOF_IPC4),)
 snd-sof-objs += ipc4.o ipc4-loader.o ipc4-topology.o ipc4-control.o ipc4-pcm.o\
-		ipc4-mtrace.o ipc4-telemetry.o
+		ipc4-mtrace.o ipc4-telemetry.o ipc4-telemetry2.o
```
what is the 2 in telemetry2?

Does this invalidate or exist concurrently with ipc4-telemetry?

jsarha replied:
Tried to answer to this question here: https://github.com/thesofproject/sof/pull/9084#discussion_r1587696123

jsarha commented:
Updated the license sections and added one something more to the commit message. But the number one question still is, what to call this debug window slot, if not telemetry2?

jsarha commented:
Closing this, as there is replacement PR here: https://github.com/thesofproject/linux/pull/5154
